### ISSUE TITLE
core npm module / package

### ISSUE THREAD

**Issue by felixfaire** [gkjohnson/closed-chain-ik-js#50]
Amazing work.

It would be fantastic if the core parts of this lib could be abstracted into an npm package. From what I can see this would only need gl-matrix and linear-solve as dependancies which would also help reduce bundle size for projects not using three.js or needing model loading etc.

**Comment by gkjohnson** [comment]
Thanks! There are still some things I would like to fix if I were to publish to npm but it can be installed from github by adding a [Github URL to the package.json](https://docs.npmjs.com/cli/v6/configuring-npm/package-json#github-urls) which is what I'd prefer at the moment:

```json
"closed-chain-ik-js": "gkjohnson/closed-chain-ik-js#v0.0.1"
```

Regarding the dependencies and making the core usable without three.js the urdf-loader and three.js dependencies can both be moved to `peerDependencies` and an `index.js` file can be added to the `src/worker/` and `src/core/` folders so they can be used without three:

```js
import { ... } from 'closed-chain-ik-js/src/core';
import { ... } from 'closed-chain-ik-js/src/worker';
```

If you want to make a PR to address these things I'd be happy to take it!

**Comment by felixfaire** [comment]
Thanks for your swift response, this sounds like a good solution for the meantime. I'm usually a native developer so forgive my relatively primitive npm knowledge.

**Comment by gkjohnson** [comment]
> though it may need vetting for best npm practices etc

Definitely happy to give feedback and help get a PR in shape if you wind up making one!

**Comment by felixfaire** [comment]
K great, incoming.

**Comment by felixfaire** [comment]
#53 Here is a first pass (though naive to npm requirements)

**Comment by felixfaire** [comment]
Sorry to bother again, but I was unable to get the package to install via the github url. Some things I also tried:

in deps
"closed-chain-ik": "gkjohnson/closed-chain-ik-js#v0.0.1"`
`"closed-chain-ik": "gkjohnson/closed-chain-ik-js#main"`
`"closed-chain-ik": "git:https://github.com/gkjohnson/closed-chain-ik-js.git#v0.0.1"`
`"closed-chain-ik-js": "git:https://github.com/gkjohnson/closed-chain-ik-js.git#v0.0.1"`
etc.
js including with 
`require('closed-chain-ik');`
``require('closed-chain-ik-js');`
``require('closed-chain-ik-js/src/core');`
etc

and various permutations. (the package name on github is 'closed-chain-ik' not 'closed-chain-ik-js' which is why i tried this but might be wrong). npm v 6.4.1, node 10.15.3. npm install warns me about the 2 peer dependancies so it must have found the repo but it never pulls the repo into the packages (causing "Can't walk dependency graph: Cannot find module 'closed-chain-ik'" on build).

I can stick with the local package for the meantime but i'd like to understand why none of this seems to work if you have any pointers as to what i'm doing wrong. Does the repo need changes to work as an npm package?

**Comment by felixfaire** [comment]
scouring through pages of google / github someone suggested deleting package-lock.json before npm install and that worked...

**Comment by gkjohnson** [comment]
Yeah the package-lock can be finnicky sometimes. Also `closed-chain-ik-js#^0.0.1` will point to [this release tag](https://github.com/gkjohnson/closed-chain-ik-js/releases/tag/v0.0.1) which doesn't have the changes I just merged from you in it. Pulling from `main` will pull from the tip of the main branch which does.

If you continue to have issues let me know and I'll make a new release with a new version number.

> It seems building the files in the package (handling import etc) doesnt work by default with my browserify setup leading to parse errors for 'import' so i'll save that for another day.

I'm less familiar with browserify (I've been using webpack, rollup, and parcel) but I don't believe it supports imports out of the box. You'll probably need to use something like Babel to transform imports (such is the modern js way...)

But glad you got it pulled! I'm gonna close this, then, if you've been able to get it all installed.
